### PR TITLE
[Feature] Training Module

### PR DESCRIPTION
**Pull request by YasheshSavani**
Description:

This PR will add a feature in Sense Studio to train your model from the web-server where user can do model selection, enter number of epochs, enter number of layers to finetune, and define where to save the checkpoint.

### CODE REVIEW

File: tools/sense_studio/static/main.css [hunk 1]
```diff
@@ -70,3 +70,14 @@ a {
...
 }
+
+.ui.segment.train-console {
+    overflow: auto;
+    height: 500px;
+    background-color: black;
+    color: white;
+}
+.font-style {
+    font-family: Ubuntu Mono;
```
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/static/main.css]: Thank you, I'll keep in mind :+1:

File: tools/sense_studio/train_module.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+def train_model():
+    data = request.form
+    project = data['project']
+    num_layers_to_finetune = data['layers_to_finetune']
+    path_out = data['output_folder']
+    model_name = data['model_name']
+    path = utils.lookup_project_path(project)
+    config = utils.load_project_config(path)
+
+    train_classifier = ["python tools/train_classifier.py", f"--path_in=/home/twentybn/Code/sense/dataset/{project}",
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:45]: :warning: That's why it doesn't work for me! You have hard-coded your own dataset path.

File: tools/sense_studio/train_module.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                        f"--num_layers_to_finetune={num_layers_to_finetune}",
+                        "--use_gpu" if config['use_gpu'] else "",
+                        f"--path_out={path_out}" if path_out else "",
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:48]: I was somehow expecting the output folder to be appended to `<project_path>/checkpoints`.
We could even indicate that on the input field by using the labeled one with the path prefix: https://semantic-ui.com/elements/input.html#labeled

File: tools/sense_studio/train_module.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                        f"--model_name={model_name}" if model_name else "",
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:49]: We need to indicate `model_name` and `model_version`

File: tools/sense_studio/train_module.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                        "--overwrite"]
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:50]: If the project is temporal, we need to include the `--temporal_training` flag and the `--path_annotations` parameters

Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:50]: You are not passing the number of `epochs` to the script

File: tools/sense_studio/train_module.py (hunk 2 above)
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:45]: The path to the `train_classifier` script should also be indicated system-independent, i.e. using `os.path.join`

File: tools/sense_studio/train_module.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    train_classifier = ' '.join(train_classifier)
+    train_classifier = shlex.split(train_classifier)
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:53]: Why exactly is `shlex` needed here?

File: tools/sense_studio/train_module.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    global PROCESS
+    PROCESS = subprocess.Popen(train_classifier, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
+
+    def generate():
+        while True:
+            output = PROCESS.stdout.readline()
+            if output == b'' and PROCESS.poll() is not None:
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:61]: Can you explain that end condition?

File: tools/sense_studio/train_module.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                        f"--model_name={model_name}" if model_name else "",
+                        "--overwrite"]
+
+    train_classifier = ' '.join(train_classifier)
+    train_classifier = shlex.split(train_classifier)
+
+    global PROCESS
+    PROCESS = subprocess.Popen(train_classifier, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
+
+    def generate():
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:58]: Please use a specific name for this function

File: tools/sense_studio/train_module.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            output = PROCESS.stdout.readline()
+            if output == b'' and PROCESS.poll() is not None:
+                PROCESS.terminate()
+                break
+            if output:
+                time.sleep(0.1)
+                yield output.decode().strip() + '\n'
+
+    return flask.Response(stream_with_context(stream_template('train.html', project=data['project'], is_disabled=True,
+                                                              models=utils.BACKBONE_MODELS, logs=generate())))
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:69]: Is it somehow possible to do this without a page refresh? Currently, all the input fields are reset when the training starts, so you don't see the parameters that you entered, and the URL also changes, even though it looks like the same page.

File: tools/sense_studio/train_module.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+    global PROCESS
+    if PROCESS:
+        PROCESS.terminate()
+        PROCESS = None
+        log = "Training Cancelled."
+    else:
+        log = "No Training Process Running to Terminate."
+    return render_template('train.html', project=data['project'], models=utils.BACKBONE_MODELS, is_disabled=False,
+                           logs=[log])
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:84]: If we use a `redirect`, the URL at least doesn't show `training/cancel-training`

File: tools/sense_studio/train_module.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    train_classifier = shlex.split(train_classifier)
+
+    global PROCESS
+    PROCESS = subprocess.Popen(train_classifier, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
+
+    def generate():
+        while True:
+            output = PROCESS.stdout.readline()
+            if output == b'' and PROCESS.poll() is not None:
+                PROCESS.terminate()
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:62]: Shouldn't we also set `PROCESS = None` in this case?

File: tools/sense_studio/train_module.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                        "--use_gpu" if config['use_gpu'] else "",
+                        f"--path_out={path_out}" if path_out else "",
+                        f"--model_name={model_name}" if model_name else "",
+                        "--overwrite"]
+
+    train_classifier = ' '.join(train_classifier)
+    train_classifier = shlex.split(train_classifier)
+
+    global PROCESS
+    PROCESS = subprocess.Popen(train_classifier, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:56]: What happens if the user leaves the page, before training has ended and without clicking `Cancel`? Will there just be an invisible long-running process in the backend?

Can we maybe indicate on the UI that there is currently a training in progress so the user can go there to cancel it?

File: tools/sense_studio/train_module.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+from flask import render_template
+from flask import request
+from flask import current_app
+from flask import stream_with_context
+
+from tools.sense_studio import utils
+
+train_bp = Blueprint('train_bp', __name__)
+
+PROCESS = None
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:18]: How about `training_process`?

File: tools/sense_studio/train_module.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    global PROCESS
+    PROCESS = subprocess.Popen(train_classifier, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
+
+    def generate():
+        while True:
+            output = PROCESS.stdout.readline()
+            if output == b'' and PROCESS.poll() is not None:
+                PROCESS.terminate()
+                break
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:63]: When training is done, the `Train Model` button is still disabled

File: tools/sense_studio/templates/train.html [hunk 15]
```diff
@@ -0,0 +1,63 @@
+{% extends 'skeleton.html' %}
```
Comment (corneliusboehm) [inline tools/sense_studio/templates/train.html:1]: I would suggest `training.html`

File: tools/sense_studio/templates/train.html [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+      <button class="ui button" type="submit" tabindex="1" formaction="{{ url_for('.cancel_training') }}">Cancel</button>
+    </form>
+    <div class="ui segment train-console">
+        {% if is_disabled %}
+            <h4 class="font-style">Training Model...</h4>
+        {% endif %}
+        <h4 class="font-style">
+            {% for log in logs %}
+                {{ log }} <br />
+            {% endfor %}
```
Comment (corneliusboehm) [inline tools/sense_studio/templates/train.html:59]: Is it possible to make this area auto-scroll to the most recent line?

File: tools/sense_studio/train_module.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    project = urllib.parse.unquote(project)
+    return render_template('train.html', project=project, models=utils.BACKBONE_MODELS, is_disabled=False)
+
+
+def stream_template(template_name, **context):
+    # Ref: https://flask.palletsprojects.com/en/1.1.x/patterns/streaming/
+    current_app.update_template_context(context)
+    t = current_app.jinja_env.get_template(template_name)
+    rv = t.stream(context)
+    return rv
```
Comment (corneliusboehm) [inline tools/sense_studio/train_module.py:32]: It looks like because of this streaming mechanism, the footer of the page is only rendered when the training is done. I'm not sure yet, if this streaming functionality is the best solution for our case, because it is originally intended for splitting up an "enormous amount of data", not for updating delayed information.

File: tools/sense_studio/static/main.js [hunk 18]
```diff
@@ -251,3 +253,17 @@ function toggleMakeProjectTemporal(path) {
         $('.temporal').hide();
     }
 }
+
+// TODO: Not Working Properly (Needs investigation)
+function disableTrainModel() {
+    let trainModel = document.getElementById('trainModel');
+    trainModel.setAttribute('disabled', 'disabled');
+    $('.train-console').show();
+}
```
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/static/main.js:262]: did you see my new approach for this of sending disabled from backend as the behaviour will be fixed?

**Comment by corneliusboehm** [comment]
Lastly, I think I would still like to extract the whole training script into a python function that we can then execute via `threading` or `multiprocessing`. This way, we're safer against changes to that script and don't rely on building a big sequence of command line arguments.

File: tools/sense_studio/templates/train.html (hunk 16 above)
Comment (YasheshSavani) [reply tools/sense_studio/templates/train.html:59]: Finding a way on that already have in mind

File: tools/sense_studio/train_module.py (hunk 14 above)
Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:63]: Have an idea on that using the new approach of sending `disabled` from backend

File: tools/sense_studio/templates/train.html [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+                </div>
+            <div class="field train">
+                <label>Number of layers to finetune</label>
+                <input type="number" name="layers_to_finetune" min="5" value="5">
+            </div>
+        </div>
+        <div class="two fields wide">
+            <div class="field train">
+                <label>Epochs</label>
+                <input type="number" name="epochs" min="80" value="80">
```
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/templates/train.html:42]: The default cases are still TBD and to be updated in the ticket. Right now, it is based on default values given in the `train_classifier` script itself.

File: tools/sense_studio/train_module.py (hunk 2 above)
Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:45]: I forgot to remove the before pushing it up, sorry.

File: tools/sense_studio/train_module.py (hunk 7 above)
Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:61]: Here, the readline() will give you an encoded string in return, and `PROCESS.poll()` returns `None` if the process is still running, if the process is completed then it will return anything like `-1` or `0`. 
So here this `and` between them helps if there is an empty line in the console log but the processing is not complete yet it will still going on.

File: tools/sense_studio/train_module.py (hunk 12 above)
Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:56]: I'll investigate on that but afaik if the user leaves that page the training process automatically gets cancelled. I have felt that but still will investigate.

File: tools/sense_studio/train_module.py (hunk 17 above)
Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:32]: I'll investigate more on `multiprocessing` and `threading` if we can do that

Comment (corneliusboehm) [reply tools/sense_studio/train_module.py:32]: Here is an example using `EventSource`, which I think should work well for us: https://stackoverflow.com/a/13388915/6611214

File: tools/sense_studio/static/main.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (corneliusboehm) [reply tools/sense_studio/static/main.js:262]: Yes, so can this code be removed?

File: tools/sense_studio/train_module.py (hunk 7 above)
Comment (corneliusboehm) [reply tools/sense_studio/train_module.py:61]: So wouldn't it be enough to only check `PROCESS.poll() is not None`, or is there a case where `poll` returns `None` but `readline` returned a non-empty string?

Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:61]: Yes, there is a case. It will happen when processing is done but the console logs are being saved in `stdout` buffer from which we are reading line one by one for display. So, here processing and reading from buffer are independent processes, so it can happen that process might be complete but we still have some lines left to read from stdout.

File: tools/sense_studio/static/main.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/static/main.js:262]: Yes

File: tools/sense_studio/train_module.py (hunk 17 above)
Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:32]: I referred the same link and you can see that after this `EventSource` approach the user  has  also posted `stream_template` approach.

Comment (corneliusboehm) [reply tools/sense_studio/train_module.py:32]: Yes, but is there an advantage to the `stream_template` approach?

Comment (YasheshSavani) [reply tools/sense_studio/train_module.py:32]: I have not tried `EventSource` in the first place and `stream_template` is provided by Flask itself. That's why I thought of using it as flask can handle its own thing properly. But there is no problem in investigating new thing which makes the feature as we want.

File: tools/sense_studio/templates/train.html [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+                            {% endfor %}
+                        </div>
+                    {% else %}
+                       <div class="default text">No models to select</div>
+                    {% endif %}
+                    </div>
+                </div>
+            <div class="field train">
+                <label>Number of layers to finetune</label>
+                <input type="number" name="layers_to_finetune" min="5" value="5">
```
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/templates/train.html]: Max is still TBD

File: tools/sense_studio/utils.py [hunk 21]
```diff
@@ -18,6 +19,8 @@
     ModelConfig('StridedInflatedMobileNetV2', 'lite', []),
 ]
 
+BACKBONE_MODELS = list(MODELS.keys())
```
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/utils.py:22]: Or we can a have a method which creates a dictionary of available model on the system, so we don't have to check before running script that the model is available or not.

**Comment by corneliusboehm** [comment]
This PR is completely contained in #142. We can test it now over there and apply potential final fixes.
